### PR TITLE
Consolidated Security Updates for 2.0.12 - develop

### PR DESCRIPTION
**Pull request by nickjjzhao**
Consolidated Security Updates for 2.0.12

Apply three-strikes rule to all transaction failures
Apply unconditional subjective CPU check along with some additional logging
Provide options to enable subjective CPU billing for P2P and API transactions ,and provide an option to disable it for individual accounts

v2.0.x - https://github.com/EOSIO/eos/pull/10264

Co-Authored-By:Kevin Heifner heifnerk@objectcomputing.com
Co-Authored-By:Lin Huang lin.huang@block.one

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [x] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

### CODE REVIEW

File: plugins/producer_plugin/producer_plugin.cpp [hunk 1]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             if (std::holds_alternative<fc::exception_ptr>(response)) {
                _transaction_ack_channel.publish(priority::low, std::pair<fc::exception_ptr, transaction_metadata_ptr>(std::get<fc::exception_ptr>(response), trx));
                if (_pending_block_mode == pending_block_mode::producing) {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid} : ${why} ",
+                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:493]: Remove the `_trx_successful_trace_log` logging here. In develop these were separated.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                        ("a", trx->packed_trx()->get_transaction().first_authorizer())
                         ("why",std::get<fc::exception_ptr>(response)->what()));
-
-                  if (_trx_trace_failure_log.is_enabled(fc::log_level::debug)) {
-                       auto entire_trace = std::get<fc::exception_ptr>(response);
-                       fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${entire_trace} ",
-                               ("block_num", chain.head_block_num() + 1)
-                               ("prod", get_pending_block_producer())
-                               ("entire_trace", entire_trace));
-                  }
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:506]: This should not have been removed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 3]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
                } else {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
+                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:506]: Remove the `_trx_successful_trace_log` logging here. In develop these were separated.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 4]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                          ("why",std::get<fc::exception_ptr>(response)->what()));
+                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                          ("txid", trx->id())
+                          ("a", trx->packed_trx()->get_transaction().first_authorizer())
                           ("why",std::get<fc::exception_ptr>(response)->what()));
-
-                  if (_trx_trace_failure_log.is_enabled(fc::log_level::debug)) {
-                       auto entire_trace = std::get<fc::exception_ptr>(response);
-                       fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${entire_trace} ", ("entire_trace", entire_trace));
-                  }
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:515]: This should not have been removed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 5]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
                           ("prod", get_pending_block_producer())
-                          ("txid", trx->id()));
-
-                  if (_trx_trace_success_log.is_enabled(fc::log_level::debug)) {
-                       auto entire_trace = chain_plug->get_entire_trx_trace(std::get<transaction_trace_ptr>(response) );
-                       fc_dlog(_trx_trace_success_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is ACCEPTING tx: ${entire_trace}",
-                               ("block_num", chain.head_block_num() + 1)
-                               ("prod", get_pending_block_producer())
-                               ("entire_trace", entire_trace));
-                  }
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:531]: This should not have been removed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 6]
```diff
@@ -491,52 +490,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                          ("txid", trx->id())
+                          ("a", trx->packed_trx()->get_transaction().first_authorizer()));
                } else {
-                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is ACCEPTING tx: ${txid}",
-                          ("txid", trx->id()));
-
-                  if (_trx_trace_success_log.is_enabled(fc::log_level::debug)) {
-                       auto entire_trace = chain_plug->get_entire_trx_trace(std::get<transaction_trace_ptr>(response) );
-                       fc_dlog(_trx_trace_success_log, "[TRX_TRACE] Speculative execution is ACCEPTING tx: ${entire_trace}", ("entire_trace", entire_trace));
-                  }
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:539]: This should not have been removed.

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/producer_plugin/producer_plugin.cpp (hunk 6 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:539]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 5 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:531]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 4 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:515]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 3 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 2 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:506]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 1 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 7]
```diff
@@ -456,15 +458,12 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                     fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                            ("txid", trx->id())("a",trx->get_transaction().first_authorizer())("why",ex->what()));
+                     fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                            ("txid", trx->id())("a",trx->get_transaction().first_authorizer())("why",ex->what()));
                      next(ex);
-
-                    if (_trx_trace_failure_log.is_enabled(fc::log_level::debug)) {
-                        auto entire_trx = self->chain_plug->get_entire_trx(trx->get_transaction());
-                        fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${entire_trx}", ("entire_trx", entire_trx));
-                    }
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:467]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 8]
```diff
@@ -456,15 +458,12 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             if( future.valid() ) {
                future.wait();
                app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}, trx{std::move(trx)}]() mutable {
-                  auto exception_handler = [self, &next, trx{std::move(trx)}](fc::exception_ptr ex) {
-                    fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
-                            ("txid", trx->id())("why",ex->what()));
+                  auto exception_handler = [&next, trx{std::move(trx)}](fc::exception_ptr ex) {
+                     fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                            ("txid", trx->id())("a",trx->get_transaction().first_authorizer())("why",ex->what()));
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 9]
```diff
@@ -457,14 +459,14 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
                future.wait();
                app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}, trx{std::move(trx)}]() mutable {
                   auto exception_handler = [self, &next, trx{std::move(trx)}](fc::exception_ptr ex) {
-                    fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
-                            ("txid", trx->id())("why",ex->what()));
+                     fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:462]: We use 3 spaces mostly in our code base.

**Review by heifner — APPROVED** [review]

**Pull request merged**